File: server/src/api/csv.ts
import type { PaymentSession } from "../domain/session.js";

/** Serialize sessions to CSV for export (AC-8.2). */
export function sessionsToCsv(sessions: PaymentSession[]): string {
  const header = [
    "id",
    "vs",
    "amount",
    "currency",
    "status",
    "overpaid",
    "createdAt",
    "expiresAt",
    "paidAt",
    "matchedTxId",
    "note",
  ];
  const rows = sessions.map((s) =>
    [
      s.id,
      s.vs,
      s.amount.toFixed(2),
      s.currency,
      s.status,
      String(s.overpaid),
      s.createdAt.toISOString(),
      s.expiresAt.toISOString(),
      s.paidAt ? s.paidAt.toISOString() : "",
      s.matchedTxId ?? "",
      s.note ?? "",
    ]
      .map(csvCell)
      .join(","),
  );
  return [header.join(","), ...rows].join("\r\n");
}

function csvCell(value: string): string {
  if (/[",\r\n]/.test(value)) {
    return `"${value.replace(/"/g, '""')}"`;
  }
  return value;
}

File: android/app/src/main/java/cz/qrplatba/api/Csv.kt
package cz.qrplatba.api

import cz.qrplatba.domain.Iso
import cz.qrplatba.domain.PaymentSession
import cz.qrplatba.domain.formatAmount2dp

/** Serialize sessions to CSV for export. Matches the Node reference column order. */
object Csv {
    private val HEADER = listOf(
        "id", "vs", "amount", "currency", "status", "overpaid",
        "createdAt", "expiresAt", "paidAt", "matchedTxId", "note",
    )

    fun sessionsToCsv(sessions: List<PaymentSession>): String {
        val rows = sessions.map { s ->
            listOf(
                s.id,
                s.vs,
                formatAmount2dp(s.amount),
                s.currency,
                s.status.name,
                s.overpaid.toString(),
                Iso.format(s.createdAt),
                Iso.format(s.expiresAt),
                s.paidAt?.let { Iso.format(it) } ?: "",
                s.matchedTxId ?: "",
                s.note ?: "",
            ).joinToString(",") { cell(it) }
        }
        return (listOf(HEADER.joinToString(",")) + rows).joinToString("\r\n")
    }

    private fun cell(value: String): String =
        if (Regex("[\",\\r\\n]").containsMatchIn(value)) {
            "\"" + value.replace("\"", "\"\"") + "\""
        } else value
}
